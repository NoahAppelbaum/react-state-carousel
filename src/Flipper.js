import { useState } from "react";
import Coin from "./Coin";


/** Flipper: Keep track of coinflips.
 *
 * State -
 * isHeads: boolean (true/false)
 * headCount: number
 * tailCount: number
 *
 * Flipper -> Coin
 */
function Flipper() {
  const [isHeads, setIsHeads] = useState();
  const [headCount, setHeadCount] = useState(0);
  const [tailCount, setTailCount] = useState(0);

  const total = headCount + tailCount;

  function flipCoin() {
    const num = Math.random();

    if (num > 0.5) {
      setIsHeads(true);
      setHeadCount(headCount + 1);
    } else {
      setIsHeads(false);
      setTailCount(tailCount + 1);
    }
  }

  return (
    <div className="Flipper">
      <h1>Let's flip a coin!</h1>
      { total > 0 && <Coin isHeads={isHeads} />}
      <button onClick={flipCoin}>FLIP MEEEEEE</button>

      <p>Out of {total} flips, there have been {headCount} heads and {tailCount} tails.</p>

    </div>
  );
}

export default Flipper;
