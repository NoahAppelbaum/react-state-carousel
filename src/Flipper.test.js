import Flipper from "./Flipper";
import { render, fireEvent } from "@testing-library/react";

Math.random = jest.fn();

it("renders without crashing", function () {
  render(<Flipper />);
});

it("should match snapshot no coin", function () {
  const { container, debug } =
    render(<Flipper />);

  expect(container).toMatchSnapshot();
});

it("should match snapshot, heads coin", function () {
  Math.random.mockReturnValue(.9);
  const { container, debug } =
    render(<Flipper />);

  const button = container.querySelector("button");
  fireEvent.click(button);

  expect(container).toMatchSnapshot();
});

it("should count totals and display correct images", function () {
  Math.random
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);

  const { container, debug } = render(<Flipper />);

  expect(container).toContainHTML(
    "Out of 0 flips, there have been 0 heads and 0 tails."
  );

  const button = container.querySelector("button");
  fireEvent.click(button);

  expect(container).toContainHTML(
    "Out of 1 flips, there have been 0 heads and 1 tails."
  );
  expect(
    container.querySelector('img[alt="tails"]')
  ).toBeInTheDocument();


  fireEvent.click(button);

  expect(container).toContainHTML(
    "Out of 2 flips, there have been 1 heads and 1 tails."
  );
  expect(
    container.querySelector('img[alt="heads"]')
  ).toBeInTheDocument();

});
