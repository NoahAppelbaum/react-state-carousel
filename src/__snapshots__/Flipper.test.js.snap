// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot no coin 1`] = `
<div>
  <div
    class="Flipper"
  >
    <h1>
      Let's flip a coin!
    </h1>
    <button>
      FLIP MEEEEEE
    </button>
    <p>
      Out of 
      0
       flips, there have been 
      0
       heads and 
      0
       tails.
    </p>
  </div>
</div>
`;

exports[`should match snapshot, heads coin 1`] = `
<div>
  <div
    class="Flipper"
  >
    <h1>
      Let's flip a coin!
    </h1>
    <div
      class="Coin"
    >
      <img
        alt="heads"
        src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/33/2010_cent_obverse.jpg/240px-2010_cent_obverse.jpg"
      />
    </div>
    <button>
      FLIP MEEEEEE
    </button>
    <p>
      Out of 
      1
       flips, there have been 
      1
       heads and 
      0
       tails.
    </p>
  </div>
</div>
`;
